REACT_EVENT doesn't support events with 0 params

### Environment
1. `react-native -v`:
react-native-cli: 2.0.1
react-native: 0.61.5
2. `react-native run-windows --info`:
 System:
    OS: Windows 10 10.0.18363
    CPU: (12) x64 Intel(R) Xeon(R) W-2133 CPU @ 3.60GHz
    Memory: 14.88 GB / 31.73 GB
  Binaries:
    Node: 10.15.1 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.0 - C:\Program Files (x86)\Yarn\bin\yarn.CMD
    npm: 6.7.0 - C:\Program Files\nodejs\npm.CMD
    Watchman: 4.9.4 - C:\Users\bhamlin\bin\watchman.EXE
  npmPackages:
    react: 16.9.0 => 16.9.0
    react-native: ^0.61.5 => 0.61.5
    react-native-windows: 0.61.0-beta.44 => 0.61.0-beta.44
  Installed UWP SDKs:
    10.0.17763.0
    10.0.18362.0

### Steps to Reproduce

Within a native module, declare an event with 0 params:

```
REACT_EVENT(MyEvent, L"myEvent")
std::function<void()> MyEvent;
```

### Expected Behavior

Compiles and I can fire my event with `MyEvent()`

### Actual Behavior
```
Microsoft.ReactNative.Cxx\NativeModules.h(802,66): error C2027: use of undefined type 'winrt::Microsoft::ReactNative::ModuleEventFieldInfo<TField>'
  with
 [
     TField=std::function<void (void)> MyModule::*
 ] (compiling source file MyModule.cpp)
```
Looks like a simple oversight, unless there's some valid reason I'm unaware of

--- a/change/react-native-windows-2020-03-25-20-43-02-FixReactEvent.json
+++ b/change/react-native-windows-2020-03-25-20-43-02-FixReactEvent.json
@@ -1,0 +1,8 @@
+{
+  "type": "prerelease",
+  "comment": "Allow NativeModules events to have 0..n arguments",
+  "packageName": "react-native-windows",
+  "email": "vmorozov@microsoft.com",
+  "dependentChangeType": "patch",
+  "date": "2020-03-26T03:43:02.397Z"
+}

--- a/vnext/Microsoft.ReactNative.Cxx.UnitTests/NativeModuleTest.cpp
+++ b/vnext/Microsoft.ReactNative.Cxx.UnitTests/NativeModuleTest.cpp
@@ -543,6 +543,14 @@ struct SimpleNativeModule {
   REACT_EVENT(OnIntEvent)
   std::function<void(int)> OnIntEvent;
 
+  // An event without arguments
+  REACT_EVENT(OnNoArgEvent)
+  std::function<void()> OnNoArgEvent;
+
+  // An event with two arguments
+  REACT_EVENT(OnTwoArgsEvent)
+  std::function<void(Point const &, Point const &)> OnTwoArgsEvent;
+
   // Specify event name different from the field name.
   REACT_EVENT(OnPointEvent, L"onPointEvent")
   std::function<void(Point const &)> OnPointEvent;
@@ -567,6 +575,10 @@ struct SimpleNativeModule {
   // Use two arguments. Specify JS function name different from the field name.
   REACT_FUNCTION(JSLineFunction, L"lineFunc")
   std::function<void(Point const &, Point const &)> JSLineFunction;
+
+  // Use no arguments.
+  REACT_FUNCTION(NoArgFunction)
+  std::function<void()> NoArgFunction;
 
   // By default we use the module name from REACT_MODULE which is by default the struct name.
   // Here we specify module name local for this function.
@@ -1327,22 +1339,51 @@ TEST_CLASS (NativeModuleTest) {
 
   TEST_METHOD(TestEvent_IntEventField) {
     bool eventRaised = false;
-    m_builderMock.ExpectEvent(L"RCTDeviceEventEmitter", L"OnIntEvent", [&eventRaised](JSValue const &arg) noexcept {
-      TestCheck(arg == 42);
-      eventRaised = true;
-    });
+    m_builderMock.ExpectEvent(
+        L"RCTDeviceEventEmitter", L"OnIntEvent", [&eventRaised](JSValueArray const &args) noexcept {
+          TestCheck(args[0] == 42);
+          eventRaised = true;
+        });
 
     m_module->OnIntEvent(42);
     TestCheck(eventRaised);
   }
 
+  TEST_METHOD(TestEvent_OnNoArgEventField) {
+    bool eventRaised = false;
+    m_builderMock.ExpectEvent(
+        L"RCTDeviceEventEmitter", L"OnNoArgEvent", [&eventRaised](JSValueArray const &args) noexcept {
+          TestCheckEqual(0, args.size());
+          eventRaised = true;
+        });
+
+    m_module->OnNoArgEvent();
+    TestCheck(eventRaised);
+  }
+
+  TEST_METHOD(TestEvent_TwoArgsEventField) {
+    bool eventRaised = false;
+    m_builderMock.ExpectEvent(
+        L"RCTDeviceEventEmitter", L"OnTwoArgsEvent", [&eventRaised](JSValueArray const &args) noexcept {
+          TestCheckEqual(4, args[0]["X"]);
+          TestCheckEqual(2, args[0]["Y"]);
+          TestCheckEqual(12, args[1]["X"]);
+          TestCheckEqual(18, args[1]["Y"]);
+          eventRaised = true;
+        });
+
+    m_module->OnTwoArgsEvent(Point{/*X =*/4, /*Y =*/2}, Point{/*X =*/12, /*Y =*/18});
+    TestCheck(eventRaised);
+  }
+
   TEST_METHOD(TestEvent_JSNameEventField) {
     bool eventRaised = false;
-    m_builderMock.ExpectEvent(L"RCTDeviceEventEmitter", L"onPointEvent", [&eventRaised](JSValue const &arg) noexcept {
-      TestCheck(arg["X"] == 4);
-      TestCheck(arg["Y"] == 2);
-      eventRaised = true;
-    });
+    m_builderMock.ExpectEvent(
+        L"RCTDeviceEventEmitter", L"onPointEvent", [&eventRaised](JSValueArray const &args) noexcept {
+          TestCheck(args[0]["X"] == 4);
+          TestCheck(args[0]["Y"] == 2);
+          eventRaised = true;
+        });
 
     m_module->OnPointEvent(Point{/*X =*/4, /*Y =*/2});
     TestCheck(eventRaised == true);
@@ -1350,8 +1391,8 @@ TEST_CLASS (NativeModuleTest) {
 
   TEST_METHOD(TestEvent_JSEventEmitterEventField) {
     bool eventRaised = false;
-    m_builderMock.ExpectEvent(L"MyEventEmitter", L"onStringEvent", [&eventRaised](JSValue const &arg) noexcept {
-      TestCheck(arg == "Hello World!");
+    m_builderMock.ExpectEvent(L"MyEventEmitter", L"onStringEvent", [&eventRaised](JSValueArray const &args) noexcept {
+      TestCheckEqual("Hello World!", args[0]);
       eventRaised = true;
     });
 
@@ -1362,9 +1403,9 @@ TEST_CLASS (NativeModuleTest) {
   TEST_METHOD(TestEvent_JSValueObjectEventField) {
     bool eventRaised = false;
     m_builderMock.ExpectEvent(
-        L"RCTDeviceEventEmitter", L"OnJSValueEvent", ([&eventRaised](JSValue const &arg) noexcept {
-          TestCheck(arg["X"] == 4);
-          TestCheck(arg["Y"] == 2);
+        L"RCTDeviceEventEmitter", L"OnJSValueEvent", ([&eventRaised](JSValueArray const &args) noexcept {
+          TestCheck(args[0]["X"] == 4);
+          TestCheck(args[0]["Y"] == 2);
           eventRaised = true;
         }));
 
@@ -1375,11 +1416,11 @@ TEST_CLASS (NativeModuleTest) {
   TEST_METHOD(TestEvent_JSValueArrayEventField) {
     bool eventRaised = false;
     m_builderMock.ExpectEvent(
-        L"RCTDeviceEventEmitter", L"OnJSValueEvent", ([&eventRaised](JSValue const &arg) noexcept {
-          TestCheck(arg[0] == "X");
-          TestCheck(arg[1] == 4);
-          TestCheck(arg[2] == true);
-          TestCheck(arg[3]["Id"] == 42);
+        L"RCTDeviceEventEmitter", L"OnJSValueEvent", ([&eventRaised](JSValueArray const &args) noexcept {
+          TestCheck(args[0][0] == "X");
+          TestCheck(args[0][1] == 4);
+          TestCheck(args[0][2] == true);
+          TestCheck(args[0][3]["Id"] == 42);
           eventRaised = true;
         }));
 
@@ -1390,8 +1431,8 @@ TEST_CLASS (NativeModuleTest) {
   TEST_METHOD(TestEvent_JSValueArray1EventField) {
     bool eventRaised = false;
     m_builderMock.ExpectEvent(
-        L"RCTDeviceEventEmitter", L"OnJSValueEvent", ([&eventRaised](JSValue const &arg) noexcept {
-          TestCheck(arg[0] == 4);
+        L"RCTDeviceEventEmitter", L"OnJSValueEvent", ([&eventRaised](JSValueArray const &args) noexcept {
+          TestCheck(args[0][0] == 4);
           eventRaised = true;
         }));
 
@@ -1437,6 +1478,18 @@ TEST_CLASS (NativeModuleTest) {
 
     m_module->JSLineFunction(Point{/*X =*/4, /*Y =*/2}, Point{/*X =*/12, /*Y =*/18});
     TestCheck(functionCalled == true);
+  }
+
+  TEST_METHOD(TestFunction_NoArgFunctionField) {
+    bool functionCalled = false;
+    m_builderMock.ExpectFunction(
+        L"SimpleNativeModule", L"NoArgFunction", [&functionCalled](JSValueArray const &args) noexcept {
+          TestCheckEqual(0, args.size());
+          functionCalled = true;
+        });
+
+    m_module->NoArgFunction();
+    TestCheck(functionCalled);
   }
 
   TEST_METHOD(TestFunction_JSModuleNameFunctionField) {

--- a/vnext/Microsoft.ReactNative.Cxx.UnitTests/ReactModuleBuilderMock.cpp
+++ b/vnext/Microsoft.ReactNative.Cxx.UnitTests/ReactModuleBuilderMock.cpp
@@ -17,15 +17,16 @@ ReactModuleBuilderMock::ReactModuleBuilderMock() noexcept : m_reactContext{make<
 void ReactModuleBuilderMock::ExpectEvent(
     std::wstring_view eventEmitterName,
     std::wstring_view eventName,
-    Mso::Functor<void(JSValue const &)> &&checkValue) noexcept {
+    Mso::Functor<void(JSValueArray const &)> &&checkValues) noexcept {
   m_jsEventHandler = [
     expectedEventEmitterName = std::wstring{eventEmitterName},
     expectedEventName = std::wstring{eventName},
-    checkValue = std::move(checkValue)
+    checkValues = std::move(checkValues)
   ](std::wstring_view eventEmitterName, std::wstring_view eventName, JSValue const &value) noexcept {
     TestCheck(expectedEventEmitterName == eventEmitterName);
     TestCheck(expectedEventName == eventName);
-    checkValue(value);
+    TestCheck(value.Type() == JSValueType::Array);
+    checkValues(value.AsArray());
   };
 }
 
@@ -59,7 +60,9 @@ void ReactModuleBuilderMock::EmitJSEvent(
     std::wstring_view eventName,
     JSValueArgWriter const &paramsArgWriter) noexcept {
   auto writer = MakeJSValueTreeWriter();
+  writer.WriteArrayBegin();
   paramsArgWriter(writer);
+  writer.WriteArrayEnd();
   m_jsEventHandler(eventEmitterName, eventName, TakeJSValue(writer));
 }
 

--- a/vnext/Microsoft.ReactNative.Cxx.UnitTests/ReactModuleBuilderMock.h
+++ b/vnext/Microsoft.ReactNative.Cxx.UnitTests/ReactModuleBuilderMock.h
@@ -58,7 +58,7 @@ struct ReactModuleBuilderMock {
   void ExpectEvent(
       std::wstring_view eventEmitterName,
       std::wstring_view eventName,
-      Mso::Functor<void(JSValue const &)> &&checkValue) noexcept;
+      Mso::Functor<void(JSValueArray const &)> &&checkValues) noexcept;
 
   void ExpectFunction(
       std::wstring_view moduleName,

--- a/vnext/Microsoft.ReactNative.Cxx/JSValueReader.h
+++ b/vnext/Microsoft.ReactNative.Cxx/JSValueReader.h
@@ -424,14 +424,7 @@ template <class... TArgs>
 inline void ReadArgs(IJSValueReader const &reader, /*out*/ TArgs &... args) noexcept {
   // Read as many arguments as we can or return default values.
   bool success = reader.ValueType() == JSValueType::Array;
-
-  if constexpr (sizeof...(args) != 0) {
-    // To read variadic template arguments in natural order we must use them in an initializer list.
-    // TODO: can we fold expression instead?
-    [[maybe_unused]] int dummy[] = {
-        (success = success && reader.GetNextArrayItem(), args = success ? ReadValue<TArgs>(reader) : TArgs{}, 0)...};
-  }
-
+  ((success = success && reader.GetNextArrayItem(), args = success ? ReadValue<TArgs>(reader) : TArgs{}), ...);
   success = success && SkipArrayToEnd(reader);
 }
 

--- a/vnext/Microsoft.ReactNative.Cxx/JSValueWriter.h
+++ b/vnext/Microsoft.ReactNative.Cxx/JSValueWriter.h
@@ -250,10 +250,7 @@ inline void WriteProperties(IJSValueWriter const &writer, T const &value) noexce
 template <class... TArgs>
 inline void WriteArgs(IJSValueWriter const &writer, TArgs const &... args) noexcept {
   writer.WriteArrayBegin();
-  if constexpr (sizeof...(args) > 0) {
-    // To write variadic template arguments in natural order we must use them in an initializer list.
-    [[maybe_unused]] int dummy[] = {(WriteValue(writer, args), 0)...};
-  }
+  (WriteValue(writer, args), ...);
   writer.WriteArrayEnd();
 }
 

--- a/vnext/Microsoft.ReactNative.Cxx/NativeModules.h
+++ b/vnext/Microsoft.ReactNative.Cxx/NativeModules.h
@@ -680,10 +680,10 @@ struct ModuleConstantInfo<void (*)(ReactConstantProvider &) noexcept> {
 template <class TField>
 struct ModuleEventFieldInfo;
 
-template <class TModule, template <class> class TFunc, class TArg>
-struct ModuleEventFieldInfo<TFunc<void(TArg)> TModule::*> {
+template <class TModule, template <class> class TFunc, class... TArgs>
+struct ModuleEventFieldInfo<TFunc<void(TArgs...)> TModule::*> {
   using ModuleType = TModule;
-  using EventType = TFunc<void(TArg)>;
+  using EventType = TFunc<void(TArgs...)>;
   using FieldType = EventType TModule::*;
 
   static InitializerDelegate GetEventHandlerInitializer(
@@ -693,10 +693,12 @@ struct ModuleEventFieldInfo<TFunc<void(TArg)> TModule::*> {
       wchar_t const *eventEmitterName) noexcept {
     return [ module = static_cast<ModuleType *>(module), field, eventName, eventEmitterName ](
         IReactContext const &reactContext) noexcept {
-      module->*field = [ reactContext, eventEmitterName, eventName ](TArg arg) noexcept {
-        reactContext.EmitJSEvent(eventEmitterName, eventName, [&arg](IJSValueWriter const &argWriter) noexcept {
-          WriteValue(argWriter, arg);
-        });
+      module->*field = [ reactContext, eventEmitterName, eventName ](TArgs... args) noexcept {
+        reactContext.EmitJSEvent(
+            eventEmitterName, eventName, [&args...]([[maybe_unused]] IJSValueWriter const &argWriter) noexcept {
+              (void)argWriter; // [[maybe_unused]] above does not work
+              (WriteValue(argWriter, args), ...);
+            });
       };
     };
   }

--- a/vnext/Microsoft.ReactNative.Managed.UnitTests/NativeModuleTest.cs
+++ b/vnext/Microsoft.ReactNative.Managed.UnitTests/NativeModuleTest.cs
@@ -456,6 +456,14 @@ namespace Microsoft.ReactNative.Managed.UnitTests
     [ReactEvent]
     public Action<int> OnIntEvent = null;
 
+    // An event without arguments
+    [ReactEvent]
+    public Action OnNoArgEvent = null;
+
+    // An event with two arguments
+    [ReactEvent]
+    public Action<Point, Point> OnTwoArgsEvent = null;
+
     // Specify event name different from the field name.
     [ReactEvent("onPointEvent")]
     public Action<Point> OnPointEvent = null;
@@ -472,6 +480,14 @@ namespace Microsoft.ReactNative.Managed.UnitTests
     // Property that allows to emit native module events.
     [ReactEvent]
     public Action<int> OnIntEventProp { get; set; }
+
+    // An event without arguments
+    [ReactEvent]
+    public Action OnNoArgEventProp { get; set; }
+
+    // An event with two arguments
+    [ReactEvent]
+    public Action<Point, Point> OnTwoArgsEventProp { get; set; }
 
     // Specify event name different from the property name.
     [ReactEvent("onPointEventProp")]
@@ -498,6 +514,10 @@ namespace Microsoft.ReactNative.Managed.UnitTests
     [ReactFunction("lineFunc")]
     public Action<Point, Point> JSLineFunction = null;
 
+    // Use no arguments.
+    [ReactFunction]
+    public Action NoArgFunction = null;
+
     // By default we use the module name from ReactModuleAttribute which is by default the class name.
     // Here we specify module name local for this function.
     [ReactFunction("stringFunc", ModuleName = "MyModule")]
@@ -518,6 +538,10 @@ namespace Microsoft.ReactNative.Managed.UnitTests
     // Use two arguments. Specify JS function name different from the property name.
     [ReactFunction("lineFuncProp")]
     public Action<Point, Point> JSLineFunctionProp { get; set; }
+
+    // Use no arguments.
+    [ReactFunction]
+    public Action NoArgFunctionProp { get; set; }
 
     // By default we use the module name from ReactModuleAttribute which is by default the class name.
     // Here we specify module name local for this function.
@@ -1075,9 +1099,9 @@ namespace Microsoft.ReactNative.Managed.UnitTests
     public void TestEvent_IntEventField()
     {
       bool eventRaised = false;
-      m_moduleBuilderMock.ExpectEvent("RCTDeviceEventEmitter", "OnIntEvent", (JSValue arg) =>
+      m_moduleBuilderMock.ExpectEvent("RCTDeviceEventEmitter", "OnIntEvent", (IReadOnlyList<JSValue> args) =>
       {
-        Assert.AreEqual(42, arg);
+        Assert.AreEqual(42, args[0]);
         eventRaised = true;
       });
 
@@ -1086,13 +1110,44 @@ namespace Microsoft.ReactNative.Managed.UnitTests
     }
 
     [TestMethod]
+    public void TestEvent_NoArgEventField()
+    {
+      bool eventRaised = false;
+      m_moduleBuilderMock.ExpectEvent("RCTDeviceEventEmitter", "OnNoArgEvent", (IReadOnlyList<JSValue> args) =>
+      {
+        Assert.AreEqual(0, args.Count);
+        eventRaised = true;
+      });
+
+      m_module.OnNoArgEvent();
+      Assert.IsTrue(eventRaised);
+    }
+
+    [TestMethod]
+    public void TestEvent_TwoArgsEventField()
+    {
+      bool eventRaised = false;
+      m_moduleBuilderMock.ExpectEvent("RCTDeviceEventEmitter", "OnTwoArgsEvent", (IReadOnlyList<JSValue> args) =>
+      {
+        Assert.AreEqual(4, args[0]["X"]);
+        Assert.AreEqual(2, args[0]["Y"]);
+        Assert.AreEqual(12, args[1]["X"]);
+        Assert.AreEqual(18, args[1]["Y"]);
+        eventRaised = true;
+      });
+
+      m_module.OnTwoArgsEvent(new Point { X = 4, Y = 2 }, new Point { X = 12, Y = 18 });
+      Assert.IsTrue(eventRaised);
+    }
+
+    [TestMethod]
     public void TestEvent_JSNameEventField()
     {
       bool eventRaised = false;
-      m_moduleBuilderMock.ExpectEvent("RCTDeviceEventEmitter", "onPointEvent", (JSValue arg) =>
+      m_moduleBuilderMock.ExpectEvent("RCTDeviceEventEmitter", "onPointEvent", (IReadOnlyList<JSValue> args) =>
       {
-        Assert.AreEqual(4, arg["X"]);
-        Assert.AreEqual(2, arg["Y"]);
+        Assert.AreEqual(4, args[0]["X"]);
+        Assert.AreEqual(2, args[0]["Y"]);
         eventRaised = true;
       });
 
@@ -1104,9 +1159,9 @@ namespace Microsoft.ReactNative.Managed.UnitTests
     public void TestEvent_JSEventEmitterEventField()
     {
       bool eventRaised = false;
-      m_moduleBuilderMock.ExpectEvent("MyEventEmitter", "onStringEvent", (JSValue arg) =>
+      m_moduleBuilderMock.ExpectEvent("MyEventEmitter", "onStringEvent", (IReadOnlyList<JSValue> args) =>
       {
-        Assert.AreEqual("Hello World!", arg);
+        Assert.AreEqual("Hello World!", args[0]);
         eventRaised = true;
       });
 
@@ -1118,10 +1173,10 @@ namespace Microsoft.ReactNative.Managed.UnitTests
     public void TestEvent_JSValueObjectEventField()
     {
       bool eventRaised = false;
-      m_moduleBuilderMock.ExpectEvent("RCTDeviceEventEmitter", "OnJSValueEvent", (JSValue arg) =>
+      m_moduleBuilderMock.ExpectEvent("RCTDeviceEventEmitter", "OnJSValueEvent", (IReadOnlyList<JSValue> args) =>
       {
-        Assert.AreEqual(4, arg["X"]);
-        Assert.AreEqual(2, arg["Y"]);
+        Assert.AreEqual(4, args[0]["X"]);
+        Assert.AreEqual(2, args[0]["Y"]);
         eventRaised = true;
       });
 
@@ -1133,12 +1188,12 @@ namespace Microsoft.ReactNative.Managed.UnitTests
     public void TestEvent_JSValueArrayEventField()
     {
       bool eventRaised = false;
-      m_moduleBuilderMock.ExpectEvent("RCTDeviceEventEmitter", "OnJSValueEvent", (JSValue arg) =>
+      m_moduleBuilderMock.ExpectEvent("RCTDeviceEventEmitter", "OnJSValueEvent", (IReadOnlyList<JSValue> args) =>
       {
-        Assert.AreEqual("X", arg[0]);
-        Assert.AreEqual(4, arg[1]);
-        Assert.AreEqual(true, arg[2]);
-        Assert.AreEqual(42, arg[3]["Id"]);
+        Assert.AreEqual("X", args[0][0]);
+        Assert.AreEqual(4, args[0][1]);
+        Assert.AreEqual(true, args[0][2]);
+        Assert.AreEqual(42, args[0][3]["Id"]);
         eventRaised = true;
       });
 
@@ -1150,9 +1205,9 @@ namespace Microsoft.ReactNative.Managed.UnitTests
     public void TestEvent_JSValueArray1EventField()
     {
       bool eventRaised = false;
-      m_moduleBuilderMock.ExpectEvent("RCTDeviceEventEmitter", "OnJSValueEvent", (JSValue arg) =>
+      m_moduleBuilderMock.ExpectEvent("RCTDeviceEventEmitter", "OnJSValueEvent", (IReadOnlyList<JSValue> args) =>
       {
-        Assert.AreEqual(42, arg[0]);
+        Assert.AreEqual(42, args[0][0]);
         eventRaised = true;
       });
 
@@ -1164,9 +1219,9 @@ namespace Microsoft.ReactNative.Managed.UnitTests
     public void TestEvent_IntEventProperty()
     {
       bool eventRaised = false;
-      m_moduleBuilderMock.ExpectEvent("RCTDeviceEventEmitter", "OnIntEventProp", (JSValue arg) =>
+      m_moduleBuilderMock.ExpectEvent("RCTDeviceEventEmitter", "OnIntEventProp", (IReadOnlyList<JSValue> args) =>
       {
-        Assert.AreEqual(42, arg);
+        Assert.AreEqual(42, args[0]);
         eventRaised = true;
       });
 
@@ -1174,14 +1229,46 @@ namespace Microsoft.ReactNative.Managed.UnitTests
       Assert.IsTrue(eventRaised);
     }
 
+
+    [TestMethod]
+    public void TestEvent_NoArgEventProperty()
+    {
+      bool eventRaised = false;
+      m_moduleBuilderMock.ExpectEvent("RCTDeviceEventEmitter", "OnNoArgEventProp", (IReadOnlyList<JSValue> args) =>
+      {
+        Assert.AreEqual(0, args.Count);
+        eventRaised = true;
+      });
+
+      m_module.OnNoArgEventProp();
+      Assert.IsTrue(eventRaised);
+    }
+
+    [TestMethod]
+    public void TestEvent_TwoArgsEventProperty()
+    {
+      bool eventRaised = false;
+      m_moduleBuilderMock.ExpectEvent("RCTDeviceEventEmitter", "OnTwoArgsEventProp", (IReadOnlyList<JSValue> args) =>
+      {
+        Assert.AreEqual(4, args[0]["X"]);
+        Assert.AreEqual(2, args[0]["Y"]);
+        Assert.AreEqual(12, args[1]["X"]);
+        Assert.AreEqual(18, args[1]["Y"]);
+        eventRaised = true;
+      });
+
+      m_module.OnTwoArgsEventProp(new Point { X = 4, Y = 2 }, new Point { X = 12, Y = 18 });
+      Assert.IsTrue(eventRaised);
+    }
+
     [TestMethod]
     public void TestEvent_JSNameEventProperty()
     {
       bool eventRaised = false;
-      m_moduleBuilderMock.ExpectEvent("RCTDeviceEventEmitter", "onPointEventProp", (JSValue arg) =>
+      m_moduleBuilderMock.ExpectEvent("RCTDeviceEventEmitter", "onPointEventProp", (IReadOnlyList<JSValue> args) =>
       {
-        Assert.AreEqual(4, arg["X"]);
-        Assert.AreEqual(2, arg["Y"]);
+        Assert.AreEqual(4, args[0]["X"]);
+        Assert.AreEqual(2, args[0]["Y"]);
         eventRaised = true;
       });
 
@@ -1193,9 +1280,9 @@ namespace Microsoft.ReactNative.Managed.UnitTests
     public void TestEvent_JSEventEmitterEventProperty()
     {
       bool eventRaised = false;
-      m_moduleBuilderMock.ExpectEvent("MyEventEmitter", "onStringEventProp", (JSValue arg) =>
+      m_moduleBuilderMock.ExpectEvent("MyEventEmitter", "onStringEventProp", (IReadOnlyList<JSValue> args) =>
       {
-        Assert.AreEqual("Hello World!", arg);
+        Assert.AreEqual("Hello World!", args[0]);
         eventRaised = true;
       });
 
@@ -1207,10 +1294,10 @@ namespace Microsoft.ReactNative.Managed.UnitTests
     public void TestEvent_JSValueObjectEventProperty()
     {
       bool eventRaised = false;
-      m_moduleBuilderMock.ExpectEvent("RCTDeviceEventEmitter", "OnJSValueEventProp", (JSValue arg) =>
+      m_moduleBuilderMock.ExpectEvent("RCTDeviceEventEmitter", "OnJSValueEventProp", (IReadOnlyList<JSValue> args) =>
       {
-        Assert.AreEqual(4, arg["X"]);
-        Assert.AreEqual(2, arg["Y"]);
+        Assert.AreEqual(4, args[0]["X"]);
+        Assert.AreEqual(2, args[0]["Y"]);
         eventRaised = true;
       });
 
@@ -1222,12 +1309,12 @@ namespace Microsoft.ReactNative.Managed.UnitTests
     public void TestEvent_JSValueArrayEventProperty()
     {
       bool eventRaised = false;
-      m_moduleBuilderMock.ExpectEvent("RCTDeviceEventEmitter", "OnJSValueEventProp", (JSValue arg) =>
+      m_moduleBuilderMock.ExpectEvent("RCTDeviceEventEmitter", "OnJSValueEventProp", (IReadOnlyList<JSValue> args) =>
       {
-        Assert.AreEqual("X", arg[0]);
-        Assert.AreEqual(4, arg[1]);
-        Assert.AreEqual(true, arg[2]);
-        Assert.AreEqual(42, arg[3]["Id"]);
+        Assert.AreEqual("X", args[0][0]);
+        Assert.AreEqual(4, args[0][1]);
+        Assert.AreEqual(true, args[0][2]);
+        Assert.AreEqual(42, args[0][3]["Id"]);
         eventRaised = true;
       });
 
@@ -1239,9 +1326,9 @@ namespace Microsoft.ReactNative.Managed.UnitTests
     public void TestEvent_JSValueArray1EventProperty()
     {
       bool eventRaised = false;
-      m_moduleBuilderMock.ExpectEvent("RCTDeviceEventEmitter", "OnJSValueEventProp", (JSValue arg) =>
+      m_moduleBuilderMock.ExpectEvent("RCTDeviceEventEmitter", "OnJSValueEventProp", (IReadOnlyList<JSValue> args) =>
       {
-        Assert.AreEqual(42, arg[0]);
+        Assert.AreEqual(42, args[0][0]);
         eventRaised = true;
       });
 
@@ -1279,7 +1366,7 @@ namespace Microsoft.ReactNative.Managed.UnitTests
     }
 
     [TestMethod]
-    public void TestFunction_TwoArgFunctionField()
+    public void TestFunction_TwoArgsFunctionField()
     {
       bool functionCalled = false;
       m_moduleBuilderMock.ExpectFunction("SimpleNativeModule", "lineFunc", (IReadOnlyList<JSValue> args) =>
@@ -1292,6 +1379,20 @@ namespace Microsoft.ReactNative.Managed.UnitTests
       });
 
       m_module.JSLineFunction(new Point { X = 4, Y = 2 }, new Point { X = 12, Y = 18 });
+      Assert.IsTrue(functionCalled);
+    }
+
+    [TestMethod]
+    public void TestFunction_NoArgFunctionField()
+    {
+      bool functionCalled = false;
+      m_moduleBuilderMock.ExpectFunction("SimpleNativeModule", "NoArgFunction", (IReadOnlyList<JSValue> args) =>
+      {
+        Assert.AreEqual(0, args.Count);
+        functionCalled = true;
+      });
+
+      m_module.NoArgFunction();
       Assert.IsTrue(functionCalled);
     }
 
@@ -1371,7 +1472,7 @@ namespace Microsoft.ReactNative.Managed.UnitTests
     }
 
     [TestMethod]
-    public void TestFunction_TwoArgFunctionProperty()
+    public void TestFunction_TwoArgsFunctionProperty()
     {
       bool functionCalled = false;
       m_moduleBuilderMock.ExpectFunction("SimpleNativeModule", "lineFuncProp", (IReadOnlyList<JSValue> args) =>
@@ -1384,6 +1485,20 @@ namespace Microsoft.ReactNative.Managed.UnitTests
       });
 
       m_module.JSLineFunctionProp(new Point { X = 4, Y = 2 }, new Point { X = 12, Y = 18 });
+      Assert.IsTrue(functionCalled);
+    }
+
+    [TestMethod]
+    public void TestFunction_NoArgFunctionProperty()
+    {
+      bool functionCalled = false;
+      m_moduleBuilderMock.ExpectFunction("SimpleNativeModule", "NoArgFunctionProp", (IReadOnlyList<JSValue> args) =>
+      {
+        Assert.AreEqual(0, args.Count);
+        functionCalled = true;
+      });
+
+      m_module.NoArgFunctionProp();
       Assert.IsTrue(functionCalled);
     }
 

--- a/vnext/Microsoft.ReactNative.Managed.UnitTests/ReactModuleBuilderMock.cs
+++ b/vnext/Microsoft.ReactNative.Managed.UnitTests/ReactModuleBuilderMock.cs
@@ -226,13 +226,14 @@ namespace Microsoft.ReactNative.Managed.UnitTests
       return constantWriter.TakeValue().AsObject();
     }
 
-    public void ExpectEvent(string eventEmitterName, string eventName, Action<JSValue> checkValue)
+    public void ExpectEvent(string eventEmitterName, string eventName, Action<IReadOnlyList<JSValue>> checkValues)
     {
       m_jsEventHandler = (string actualEventEmitterName, string actualEventName, JSValue value) =>
       {
         Assert.AreEqual(eventEmitterName, actualEventEmitterName);
         Assert.AreEqual(eventName, actualEventName);
-        checkValue(value);
+        Assert.AreEqual(JSValueType.Array, value.Type);
+        checkValues(value.AsArray());
       };
     }
 
@@ -257,7 +258,9 @@ namespace Microsoft.ReactNative.Managed.UnitTests
     public void EmitJSEvent(string eventEmitterName, string eventName, JSValueArgWriter paramsArgWriter)
     {
       var writer = new JSValueTreeWriter();
+      writer.WriteArrayBegin();
       paramsArgWriter(writer);
+      writer.WriteArrayEnd();
       m_jsEventHandler(eventEmitterName, eventName, writer.TakeValue());
     }
   }

--- a/vnext/Microsoft.ReactNative.SharedManaged/ReactContextExtensions.cs
+++ b/vnext/Microsoft.ReactNative.SharedManaged/ReactContextExtensions.cs
@@ -56,9 +56,90 @@ namespace Microsoft.ReactNative.Managed
       reactContext.CallJSFunction(moduleName, methodName, (IJSValueWriter writer) => writer.WriteArgs(arg1, arg2, arg3, arg4, arg5, arg6, arg7));
     }
 
-    public static void EmitJSEvent<T>(this IReactContext reactContext, string eventEmitterName, string eventName, T arg)
+    public static void EmitJSEvent(this IReactContext reactContext, string eventEmitterName, string eventName)
     {
-      reactContext.EmitJSEvent(eventEmitterName, eventName, (IJSValueWriter writer) => writer.WriteValue(arg));
+      reactContext.EmitJSEvent(eventEmitterName, eventName, (IJSValueWriter writer) => {});
+    }
+
+    public static void EmitJSEvent<T1>(this IReactContext reactContext, string eventEmitterName, string eventName, T1 arg1)
+    {
+      reactContext.EmitJSEvent(eventEmitterName, eventName, (IJSValueWriter writer) =>
+      {
+        writer.WriteValue(arg1);
+      });
+    }
+
+    public static void EmitJSEvent<T1, T2>(this IReactContext reactContext, string eventEmitterName, string eventName, T1 arg1, T2 arg2)
+    {
+      reactContext.EmitJSEvent(eventEmitterName, eventName, (IJSValueWriter writer) =>
+      {
+        writer.WriteValue(arg1);
+        writer.WriteValue(arg2);
+      });
+    }
+
+    public static void EmitJSEvent<T1, T2, T3>(this IReactContext reactContext, string eventEmitterName, string eventName, T1 arg1, T2 arg2, T3 arg3)
+    {
+      reactContext.EmitJSEvent(eventEmitterName, eventName, (IJSValueWriter writer) =>
+      {
+        writer.WriteValue(arg1);
+        writer.WriteValue(arg2);
+        writer.WriteValue(arg3);
+      });
+    }
+
+    public static void EmitJSEvent<T1, T2, T3, T4>(this IReactContext reactContext,
+      string eventEmitterName, string eventName, T1 arg1, T2 arg2, T3 arg3, T4 arg4)
+    {
+      reactContext.EmitJSEvent(eventEmitterName, eventName, (IJSValueWriter writer) =>
+      {
+        writer.WriteValue(arg1);
+        writer.WriteValue(arg2);
+        writer.WriteValue(arg3);
+        writer.WriteValue(arg4);
+      });
+    }
+
+    public static void EmitJSEvent<T1, T2, T3, T4, T5>(this IReactContext reactContext,
+      string eventEmitterName, string eventName, T1 arg1, T2 arg2, T3 arg3, T4 arg4, T5 arg5)
+    {
+      reactContext.EmitJSEvent(eventEmitterName, eventName, (IJSValueWriter writer) =>
+      {
+        writer.WriteValue(arg1);
+        writer.WriteValue(arg2);
+        writer.WriteValue(arg3);
+        writer.WriteValue(arg4);
+        writer.WriteValue(arg5);
+      });
+    }
+
+    public static void EmitJSEvent<T1, T2, T3, T4, T5, T6>(this IReactContext reactContext,
+      string eventEmitterName, string eventName, T1 arg1, T2 arg2, T3 arg3, T4 arg4, T5 arg5, T6 arg6)
+    {
+      reactContext.EmitJSEvent(eventEmitterName, eventName, (IJSValueWriter writer) =>
+      {
+        writer.WriteValue(arg1);
+        writer.WriteValue(arg2);
+        writer.WriteValue(arg3);
+        writer.WriteValue(arg4);
+        writer.WriteValue(arg5);
+        writer.WriteValue(arg6);
+      });
+    }
+
+    public static void EmitJSEvent<T1, T2, T3, T4, T5, T6, T7>(this IReactContext reactContext,
+      string eventEmitterName, string eventName, T1 arg1, T2 arg2, T3 arg3, T4 arg4, T5 arg5, T6 arg6, T7 arg7)
+    {
+      reactContext.EmitJSEvent(eventEmitterName, eventName, (IJSValueWriter writer) =>
+      {
+        writer.WriteValue(arg1);
+        writer.WriteValue(arg2);
+        writer.WriteValue(arg3);
+        writer.WriteValue(arg4);
+        writer.WriteValue(arg5);
+        writer.WriteValue(arg6);
+        writer.WriteValue(arg7);
+      });
     }
   }
 }

--- a/vnext/Microsoft.ReactNative.SharedManaged/ReactEventInfo.cs
+++ b/vnext/Microsoft.ReactNative.SharedManaged/ReactEventInfo.cs
@@ -2,6 +2,7 @@
 // Licensed under the MIT License.
 
 using System;
+using System.Linq;
 using System.Reflection;
 using System.Threading;
 using static Microsoft.ReactNative.Managed.JSValueGenerator;
@@ -16,23 +17,23 @@ namespace Microsoft.ReactNative.Managed
     {
       var eventName = eventAttribute.EventName ?? propertyInfo.Name;
       eventEmitterName = eventAttribute.EventEmitterName ?? eventEmitterName;
-      var eventArgType = GetEventArgType(propertyInfo.PropertyType,
-        formatError: message => $"{message} Module: {propertyInfo.DeclaringType.Name}, Field: {propertyInfo.Name}");
+      var eventArgTypes = GetEventArgTypes(propertyInfo.PropertyType,
+        formatError: message => $"{message} Module: {propertyInfo.DeclaringType.Name}, Property: {propertyInfo.Name}");
       EventImpl = new Lazy<ReactEventImpl>(() =>
-        MakeEvent(propertyInfo, eventEmitterName, eventName, eventArgType), LazyThreadSafetyMode.PublicationOnly);
+        MakeEvent(propertyInfo, eventEmitterName, eventName, eventArgTypes), LazyThreadSafetyMode.PublicationOnly);
     }
 
     public ReactEventInfo(FieldInfo fieldInfo, ReactEventAttribute eventAttribute, string eventEmitterName)
     {
       var eventName = eventAttribute.EventName ?? fieldInfo.Name;
       eventEmitterName = eventAttribute.EventEmitterName ?? eventEmitterName;
-      var eventArgType = GetEventArgType(fieldInfo.FieldType,
+      var eventArgTypes = GetEventArgTypes(fieldInfo.FieldType,
         formatError: message => $"{message} Module: {fieldInfo.DeclaringType.Name}, Field: {fieldInfo.Name}");
       EventImpl = new Lazy<ReactEventImpl>(() =>
-        MakeEvent(fieldInfo, eventEmitterName, eventName, eventArgType), LazyThreadSafetyMode.PublicationOnly);
+        MakeEvent(fieldInfo, eventEmitterName, eventName, eventArgTypes), LazyThreadSafetyMode.PublicationOnly);
     }
 
-    private static Type GetEventArgType(Type eventType, Func<string, string> formatError)
+    private static Type[] GetEventArgTypes(Type eventType, Func<string, string> formatError)
     {
       if (!typeof(Delegate).IsAssignableFrom(eventType))
       {
@@ -41,25 +42,21 @@ namespace Microsoft.ReactNative.Managed
 
       MethodInfo eventDelegateMethod = eventType.GetMethod("Invoke");
       ParameterInfo[] parameters = eventDelegateMethod.GetParameters();
-      if (parameters.Length != 1)
-      {
-        throw new InvalidOperationException(formatError("React event delegate must have one parameter."));
-      }
 
-      return parameters[0].ParameterType;
+      return parameters.Select(p => p.ParameterType).ToArray();
     }
 
     static MethodInfo EmitJSEvent() =>
       typeof(IReactContext).GetMethod(nameof(IReactContext.EmitJSEvent));
 
-    private ReactEventImpl MakeEvent(PropertyInfo propertyInfo, string eventEmitterName, string eventName, Type eventArgType)
+    private ReactEventImpl MakeEvent(PropertyInfo propertyInfo, string eventEmitterName, string eventName, Type[] eventArgTypes)
     {
       // Generate code that looks like:
       //
       // (object module, IReactContext reactContext) =>
       // {
-      //   (module as MyModule).EventProperty = (ArgType arg) =>
-      //     reactContext.EmitJSEvent(eventEmitterName, eventName, arg);
+      //   (module as MyModule).EventProperty = (ArgType0 arg0, ArgType1 arg0) =>
+      //     reactContext.EmitJSEvent(eventEmitterName, eventName, arg0, arg1);
       // });
 
       return CompileLambda<ReactEventImpl>(
@@ -67,18 +64,18 @@ namespace Microsoft.ReactNative.Managed
         Parameter(typeof(IReactContext), out var reactContext),
         module.CastTo(propertyInfo.DeclaringType).SetProperty(propertyInfo,
           AutoLambda(propertyInfo.PropertyType,
-            Parameter(eventArgType, out var arg),
-            reactContext.CallExt(EmitJSEventOf(eventArgType), Constant(eventEmitterName), Constant(eventName), arg))));
+            Parameters(eventArgTypes, out var args),
+            reactContext.CallExt(EmitJSEventOf(eventArgTypes), Constant(eventEmitterName), Constant(eventName), args))));
     }
 
-    private ReactEventImpl MakeEvent(FieldInfo fieldInfo, string eventEmitterName, string eventName, Type eventArgType)
+    private ReactEventImpl MakeEvent(FieldInfo fieldInfo, string eventEmitterName, string eventName, Type[] eventArgTypes)
     {
       // Generate code that looks like:
       //
       // (object module, IReactContext reactContext) =>
       // {
-      //   (module as MyModule).EventField = (ArgType arg) =>
-      //     reactContext.EmitJSEvent(eventEmitterName, eventName, arg);
+      //   (module as MyModule).EventField = (ArgType0 arg0, ArgType1 arg0) =>
+      //     reactContext.EmitJSEvent(eventEmitterName, eventName, arg0, arg1);
       // });
 
       return CompileLambda<ReactEventImpl>(
@@ -86,8 +83,8 @@ namespace Microsoft.ReactNative.Managed
         Parameter(typeof(IReactContext), out var reactContext),
         module.CastTo(fieldInfo.DeclaringType).SetField(fieldInfo,
           AutoLambda(fieldInfo.FieldType,
-            Parameter(eventArgType, out var arg),
-            reactContext.CallExt(EmitJSEventOf(eventArgType), Constant(eventEmitterName), Constant(eventName), arg))));
+            Parameters(eventArgTypes, out var args),
+            reactContext.CallExt(EmitJSEventOf(eventArgTypes), Constant(eventEmitterName), Constant(eventName), args))));
     }
 
     public delegate void ReactEventImpl(object module, IReactContext reactContext);

